pkg/plugin/integration_test.go

```diff
@@ -351,27 +316,377 @@ func TestCreateThenUpdateCausesNoRotations(t *testing.T) {
+				for i := range cs.Properties.AgentPoolProfiles {
+					cs.Properties.AgentPoolProfiles[i].SubnetCIDR = "11.0.0.0/24"
+				}
+			},
+		},
+	}
+
+	log := logrus.NewEntry(logrus.StandardLogger())
+	ctx := context.Background()
+	cs := newTestCs()
+	az := newFakeAzureCloud(log)
+	p, _, err := setupNewCluster(ctx, log, cs, az)
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			log.Infof("--- Test: %s", tt.name)
+			beforeBlob, beforeSyncChecksum, err := getHashes(az, cs)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			tt.change(cs)
+
+			errs := p.Validate(ctx, cs, nil, false)
+			if errs != nil {
+				t.Fatal(errs)
+			}
+			perr := p.CreateOrUpdate(ctx, cs, true, getFakeDeployer(log, cs, az))
+			if perr != nil {
+				t.Fatal(perr)
+			}
+
+			afterBlob, afterSyncChecksum, err := getHashes(az, cs)
+			if err != nil {
+				t.Fatal(err)
+			}
+			rotations := getRotations(beforeBlob, afterBlob, beforeSyncChecksum, afterSyncChecksum)
+			if !reflect.DeepEqual(tt.expectRotation, rotations) {
+				t.Fatalf("rotation mismatch: expected %v, got %v", tt.expectRotation, rotations)
+			}
+		})
+	}
+}
+
+func TestHowActionsCauseRotations(t *testing.T) {
+	tests := []struct {
+		name           string
+		change         func(ctx context.Context, log *logrus.Entry, p *plugin, cs *api.OpenShiftManagedCluster, az *fakecloud.AzureCloud) error
+		expectRotation map[rotationType]bool
+		expectNodes    map[rotationType]int
+		expectCalls    []string
+	}{
+		{
+			name:           "scale up",
+			expectRotation: map[rotationType]bool{rotationMaster: false, rotationInfra: false, rotationSync: false, rotationCompute: false},
+			expectNodes:    map[rotationType]int{rotationCompute: 6, rotationInfra: 4, rotationMaster: 3},
+			change: func(ctx context.Context, log *logrus.Entry, p *plugin, cs *api.OpenShiftManagedCluster, az *fakecloud.AzureCloud) error {
+				for i, p := range cs.Properties.AgentPoolProfiles {
+					if p.Role == api.AgentPoolProfileRoleInfra {
+						cs.Properties.AgentPoolProfiles[i].Count = 4
+					}
+					if p.Role == api.AgentPoolProfileRoleCompute {
+						cs.Properties.AgentPoolProfiles[i].Count = 6
+					}
+				}
+				perr := p.CreateOrUpdate(ctx, cs, true, getFakeDeployer(log, cs, az))
+				if perr != nil {
+					return perr
+				}
+				return nil
+			},
+		},
+		{
+			name:           "rotate cluster secrets",
+			expectRotation: map[rotationType]bool{rotationMaster: true, rotationInfra: true, rotationSync: true, rotationCompute: true},
+			expectNodes:    map[rotationType]int{rotationCompute: 1, rotationInfra: 2, rotationMaster: 3},
+			change: func(ctx context.Context, log *logrus.Entry, p *plugin, cs *api.OpenShiftManagedCluster, az *fakecloud.AzureCloud) error {
+				perr := p.RotateClusterSecrets(ctx, cs, getFakeDeployer(log, cs, az))
+				if perr != nil {
+					return perr
+				}
+				return nil
+			},
+		},
+		{
+			name:           "readonly actions cause no rotations",
```

Comment:
It might be worth splitting this into separate unit tests, since if one of these actions fails for some reason, we can't actually tell which one very easily.

Reply:
I could, just trying to reduce the test time. each test takes 3-4 seconds.

Reply:
That's a bummer, but makes sense.